comb_spec_searcher/limited_strategy_db.py

```diff
@@ -0,0 +1,149 @@
+"""
+    An alternative ruledb that permits restricting the number of applications of a given
+    set of strategies to a given limit.
+"""
+from copy import deepcopy
+from itertools import combinations
+from time import time
+from typing import Dict, Set, Tuple, Type
+
+from logzero import logger
+
+from comb_spec_searcher.rule_db import RuleDB
+from comb_spec_searcher.strategies import AbstractStrategy
+
+from .exception import SpecificationNotFound
+from .tree_searcher import (
+    Node,
+    iterative_proof_tree_finder,
+    iterative_prune,
+    prune,
+    random_proof_tree,
+)
+
+LabelRule = Tuple[int, Tuple[int, ...]]
+RulesDict = Dict[int, Set[Tuple[int, ...]]]
+
+
+class LimitedStrategyRuleDB(RuleDB):
+    def __init__(
+        self,
+        strategies_to_limit: Set[Type[AbstractStrategy]],
+        limit: int,
+        mark_verified: bool,
+        minimization_time_limit: int,
+    ) -> None:
+        self.strategies_to_limit = strategies_to_limit
+        self.limit = limit
+        self.mark_verified = mark_verified
+        self.minimization_time_limit = minimization_time_limit
+        super().__init__()
+
+    def get_eqv_rules_using_strategies(self) -> Set[LabelRule]:
+        """
+        returns a set of rules, up to equivalence, that come from
+        <self.strategies_to_limit>
+        """
+        eqv_rules_using_strategies: Set[LabelRule] = set()
+        for label_rule in self:
+            rule_strat = self.rule_to_strategy[label_rule]
+            start, ends = label_rule
+            if any(isinstance(rule_strat, strat) for strat in self.strategies_to_limit):
```

Comment:
You could just write 
```python
if isinstance(rule_strat, self.strategies_to_limit):
```

Reply:
Actually, I find this less readable (I guess because of my own ignorance in not knowing this construction). I will still change it if either of you feel strongly that I should.

Reply:
I personally find it more readable, so feel semi-strongly about it, but also not willing to argue for it.

Reply:
Interesting! I've changed it now.